README.md
A Flutter Package that can validate strings

## Features

#### A flutter Package that:

* Can validate string values
* Will provide accurate results

## Additional information

* Easy to use and manage
* You just have to provide your params

## Basic Usage

```

 @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Center(
            child: Column(mainAxisAlignment: MainAxisAlignment.center, children: <Widget>[
      TextButton(
          onPressed: () {
            _validateStrings("1111", context);
          },
          child: Center(child: Text('Check 1111', style: Theme.of(context).textTheme.headlineMedium))),
      TextButton(
          onPressed: () {
            _validateStrings("1221", context);
          },
          child: Center(child: Text('Check 1221', style: Theme.of(context).textTheme.headlineMedium))),
      TextButton(
          onPressed: () {
            _validateStrings("abc@gmail.com", context);
          },
          child: Center(child: Text('Check abc@gmail.com', style: Theme.of(context).textTheme.headlineMedium)))
    ])));
  }

  void _validateStrings(String givenString, BuildContext context) {
    if (StringHelperUtils.instance.isNullOrBlank(givenString: givenString)) {
      debugPrint("Value is null or empty");
    } else if (StringHelperUtils.instance.isDouble(stringToCheck: givenString)) {
      debugPrint("Value is Double");
    } else if (StringHelperUtils.instance.isInt(stringToCheck: givenString)) {
      debugPrint("Value is Int");
    } else if (StringHelperUtils.instance.isNumeric(stringValue: givenString)) {
      debugPrint("Value is Numeric");
    } else if (StringHelperUtils.instance.isValidEmail(givenEmail: givenString)) {
      debugPrint("Valid Email");
    } else if (StringHelperUtils.instance.isValidAmountValue(amountValue: givenString, allowedMaxAmount: 12345)) {
      debugPrint("Valid Amount");
    }
  }

```

CHANGELOG.md
## 0.0.1

* Initial Release

## 0.0.2

* Example Added
* Readme updated
* Project description updated


test/string_utils_abdullah_0120_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:string_utils_abdullah_0120/string_utils_abdullah_0120.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/string_utils_abdullah.dart
import 'package:dartx/dartx_io.dart';
import 'package:email_validator/email_validator.dart';

///To check if the text field entering amount values is valid or not
///[amountValue] is required
///[baseCurrencyString] is optional(can be useful in future) but if not null or empty, will be replaced with voids
///then check if the rest amount is numeric or not
bool isValidAmountValue({required String amountValue, required double allowedMaxAmount, String baseCurrencyString = ""}) {
  bool isValid = false;
  if (amountValue.isNotEmpty) {
    String amountToCheck = getAmountValueWithoutCurrency(givenAmount: amountValue);
    if (isNumeric(stringValue: amountToCheck) && double.tryParse(amountToCheck) != null) {
      double amount = double.tryParse(amountToCheck)!;
      if (amount <= allowedMaxAmount) {
        isValid = true;
      }
    }
  }
  return isValid;
}

///Method to check if the given string only contains
///Numbers or decimals
bool isNumeric({String? stringValue}) {
  if (stringValue != null && stringValue.isNotEmpty) {
    return double.tryParse(stringValue) != null;
  }
  return false;
}

///To check if given String id Double type
///[stringToCheck] is required and can be null
bool isDouble({String? stringToCheck}) {
  return (stringToCheck ?? "").isDouble;
}

///To check if given String id Integer type
///[stringToCheck] is required and can be null
bool isInt({String? stringToCheck}) {
  return (stringToCheck ?? "").isInt;
}

/// isNull = null.isNullOrBlank; // true
/// isEmpty = ''.isNullOrBlank; // true
/// isBlank = ' '.isNullOrBlank; // true
/// isLineBreak = '\n'.isNullOrBlank; // true
/// isFoo = ' foo '.isNullOrBlank; // false
bool isNullOrBlank({String? givenString}) {
  return givenString.isNullOrBlank;
}

///To check if given email string is valid
///[givenEmail] is required
bool isValidEmail({required String givenEmail}) {
  if (isNullOrBlank(givenString: givenEmail)) {
    return false;
  }
  return EmailValidator.validate(givenEmail);
}

///Get the amount string for the BE purpose without the currency symbol
///Without the space
///Only with the numbers
String getAmountValueWithoutCurrency({required String givenAmount}) {
  String amount = "0";
  if (givenAmount.isNotEmpty) {
    if (givenAmount.contains(".")) {
      String amountLeftToDecimal = givenAmount.split(".")[0];
      String amountRightToDecimal = givenAmount.split(".")[1];

      String numericAmountLeftToDecimal = extractOnlyNumbersFromString(alphaNumericString: amountLeftToDecimal) ?? amount;
      String numericAmountRightToDecimal = extractOnlyNumbersFromString(alphaNumericString: amountRightToDecimal) ?? amount;

      amount = "$numericAmountLeftToDecimal.$numericAmountRightToDecimal";
    } else {
      amount = extractOnlyNumbersFromString(alphaNumericString: givenAmount) ?? amount;
    }
  }
  return amount;
}

///extract the numbers from the string
///[alphaNumericString] will be like 100 98.22 Ar
///The return for example would be 1009822 only numbers
///The returned result can be null
String? extractOnlyNumbersFromString({required String alphaNumericString}) {
  return alphaNumericString.isNotEmpty ? ((alphaNumericString).replaceAll(RegExp(r'\D'), '')) : null;
}


lib/string_utils_abdullah_0120.dart
import 'package:dartx/dartx_io.dart';
import 'package:email_validator/email_validator.dart';

class StringHelperUtils {
  StringHelperUtils._();
  static final instance = StringHelperUtils._();

  ///To check if the text field entering amount values is valid or not
  ///[amountValue] is required
  ///[baseCurrencyString] is optional(can be useful in future) but if not null or empty, will be replaced with voids
  ///then check if the rest amount is numeric or not
  bool isValidAmountValue({required String amountValue, required double allowedMaxAmount, String baseCurrencyString = ""}) {
    bool isValid = false;
    if (amountValue.isNotEmpty) {
      String amountToCheck = getAmountValueWithoutCurrency(givenAmount: amountValue);
      if (isNumeric(stringValue: amountToCheck) && double.tryParse(amountToCheck) != null) {
        double amount = double.tryParse(amountToCheck)!;
        if (amount <= allowedMaxAmount) {
          isValid = true;
        }
      }
    }
    return isValid;
  }

  ///Method to check if the given string only contains
  ///Numbers or decimals
  bool isNumeric({String? stringValue}) {
    if (stringValue != null && stringValue.isNotEmpty) {
      return double.tryParse(stringValue) != null;
    }
    return false;
  }

  ///To check if given String id Double type
  ///[stringToCheck] is required and can be null
  bool isDouble({String? stringToCheck}) {
    return (stringToCheck ?? "").isDouble;
  }

  ///To check if given String id Integer type
  ///[stringToCheck] is required and can be null
  bool isInt({String? stringToCheck}) {
    return (stringToCheck ?? "").isInt;
  }

  /// isNull = null.isNullOrBlank; // true
  /// isEmpty = ''.isNullOrBlank; // true
  /// isBlank = ' '.isNullOrBlank; // true
  /// isLineBreak = '\n'.isNullOrBlank; // true
  /// isFoo = ' foo '.isNullOrBlank; // false
  bool isNullOrBlank({String? givenString}) {
    return givenString.isNullOrBlank;
  }

  ///To check if given email string is valid
  ///[givenEmail] is required
  bool isValidEmail({required String givenEmail}) {
    if (isNullOrBlank(givenString: givenEmail)) {
      return false;
    }
    return EmailValidator.validate(givenEmail);
  }

  ///Get the amount string for the BE purpose without the currency symbol
  ///Without the space
  ///Only with the numbers
  String getAmountValueWithoutCurrency({required String givenAmount}) {
    String amount = "0";
    if (givenAmount.isNotEmpty) {
      if (givenAmount.contains(".")) {
        String amountLeftToDecimal = givenAmount.split(".")[0];
        String amountRightToDecimal = givenAmount.split(".")[1];

        String numericAmountLeftToDecimal = extractOnlyNumbersFromString(alphaNumericString: amountLeftToDecimal) ?? amount;
        String numericAmountRightToDecimal = extractOnlyNumbersFromString(alphaNumericString: amountRightToDecimal) ?? amount;

        amount = "$numericAmountLeftToDecimal.$numericAmountRightToDecimal";
      } else {
        amount = extractOnlyNumbersFromString(alphaNumericString: givenAmount) ?? amount;
      }
    }
    return amount;
  }

  ///extract the numbers from the string
  ///[alphaNumericString] will be like 100 98.22 Ar
  ///The return for example would be 1009822 only numbers
  ///The returned result can be null
  String? extractOnlyNumbersFromString({required String alphaNumericString}) {
    return alphaNumericString.isNotEmpty ? ((alphaNumericString).replaceAll(RegExp(r'\D'), '')) : null;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:string_utils_abdullah_0120/string_utils_abdullah_0120.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Center(
            child: Column(mainAxisAlignment: MainAxisAlignment.center, children: <Widget>[
      TextButton(
          onPressed: () {
            _validateStrings("1111", context);
          },
          child: Center(child: Text('Check 1111', style: Theme.of(context).textTheme.headlineMedium))),
      TextButton(
          onPressed: () {
            _validateStrings("1221", context);
          },
          child: Center(child: Text('Check 1221', style: Theme.of(context).textTheme.headlineMedium))),
      TextButton(
          onPressed: () {
            _validateStrings("abc@gmail.com", context);
          },
          child: Center(child: Text('Check abc@gmail.com', style: Theme.of(context).textTheme.headlineMedium)))
    ])));
  }

  void _validateStrings(String givenString, BuildContext context) {
    if (StringHelperUtils.instance.isNullOrBlank(givenString: givenString)) {
      debugPrint("Value is null or empty");
    } else if (StringHelperUtils.instance.isDouble(stringToCheck: givenString)) {
      debugPrint("Value is Double");
    } else if (StringHelperUtils.instance.isInt(stringToCheck: givenString)) {
      debugPrint("Value is Int");
    } else if (StringHelperUtils.instance.isNumeric(stringValue: givenString)) {
      debugPrint("Value is Numeric");
    } else if (StringHelperUtils.instance.isValidEmail(givenEmail: givenString)) {
      debugPrint("Valid Email");
    } else if (StringHelperUtils.instance.isValidAmountValue(amountValue: givenString, allowedMaxAmount: 12345)) {
      debugPrint("Valid Amount");
    }
  }
}
